Userland/Utilities/run-tests.cpp

```diff
@@ -0,0 +1,327 @@
+    const char* specified_test_root = nullptr;
+    String test_glob;
+    String exclude_pattern;
+    String config_file;
+
+    Core::ArgsParser args_parser;
+    args_parser.add_option(Core::ArgsParser::Option {
+        .requires_argument = true,
+        .help_string = "Show progress with OSC 9 (true, false)",
+        .long_name = "show-progress",
+        .short_name = 'p',
+        .accept_value = [&](auto* str) {
+            if (StringView { "true" } == str)
+                print_progress = true;
+            else if (StringView { "false" } == str)
+                print_progress = false;
+            else
+                return false;
+            return true;
+        },
+    });
+    args_parser.add_option(print_json, "Show results as JSON", "json", 'j');
+    args_parser.add_option(print_all_output, "Show all test output", "verbose", 'v');
+    args_parser.add_option(test_glob, "Only run tests matching the given glob", "filter", 'f', "glob");
+    args_parser.add_option(exclude_pattern, "Regular expression to use to exclude paths from being considered tests", "exclude-pattern", 'e', "pattern");
+    args_parser.add_option(config_file, "Configuration file to use", "config-file", 'c', "filename");
+    args_parser.add_positional_argument(specified_test_root, "Tests root directory", "path", Core::ArgsParser::Required::No);
+    args_parser.parse(argc, argv);
+
+    test_glob = String::formatted("*{}*", test_glob);
+
+    if (getenv("DISABLE_DBG_OUTPUT")) {
+        AK::set_debug_enabled(false);
+    }
+
+    String test_root;
+
+    if (specified_test_root) {
+        test_root = String { specified_test_root };
+    } else {
+        test_root = "/usr/Tests";
+    }
+    if (!Core::File::is_directory(test_root)) {
+        warnln("Test root is not a directory: {}", test_root);
+        return 1;
+    }
+
+    test_root = Core::File::real_path_for(test_root);
+
+    if (chdir(test_root.characters()) < 0) {
+        auto saved_errno = errno;
+        warnln("chdir failed: {}", strerror(saved_errno));
+        return 1;
+    }
+
+    auto config = config_file.is_empty() ? Core::ConfigFile::get_for_app("Tests") : Core::ConfigFile::open(config_file);
+    if (config->num_groups() == 0)
+        warnln("Empty configuration file ({}) loaded!", config_file.is_empty() ? "User config for Tests" : config_file.characters());
+
+    if (exclude_pattern.is_empty())
+        exclude_pattern = config->read_entry("Global", "NotTestsPattern", "$^"); // default is match nothing (aka match end then beginning)
+
+    Regex<PosixExtended> exclude_regex(exclude_pattern, {});
+    if (exclude_regex.parser_result.error != Error::NoError) {
+        warnln("Exclude pattern \"{}\" is invalid", exclude_pattern);
+        return 1;
+    }
+
+    TestRunner test_runner(test_root, move(exclude_regex), move(config), print_progress, print_json, print_all_output);
+    test_runner.run(test_glob);
+
+    return test_runner.counts().tests_failed;
```

Comment:
me no likey how much of this is copied
Ideally we'd just add the features this supports to the base runner and have a single `main`, but that's just _ideal_

Reply:
yeahhhh I think we'd ideally create a generic main, and make the "run a bunch of executables in a directory" runner and the JS runner just add extra stuff into those. the number of global vars in the JS test runner is terrifying though. Not sure I'm up for that level of refactor at the moment 🤔

Reply:
np, that can be done later too :^)